### PR TITLE
Desktop: Fixes #8382: add "insert*before" buttons for tables in the rich text editor

### LINKED ISSUE #8382: "Insert row before" missing since 2.11.11
"Insert row before" button is simply not there since 2.11.11

## Environment

Joplin version: 2.11.11
Platform: Windows
OS specifics: 10

## Steps to reproduce

1. click in a table, row of buttons appears below
2. "Insert row before" button is not there

### PR DESCRIPTION
**Pull request by marph91**
Fixes #8382 

Before
table_context_menu_before

After
table_context_menu_after

Should we add the "paste before" buttons, too? It would be consistent to the insert buttons, but take some more space.

### CODE REVIEW

**Comment by DeirdreYoung** [comment]
Can "Delete table" not be the very first button at the left? I have already hit it it once by accident. In fact, maybe that button is not needed at all? And yes to a "Paste before" button, please.

**Comment by laurent22** [comment]
@marph91, it looks like the buttons below are the default ones, at least according to [this demo](https://www.tiny.cloud/docs/demo/full-featured/)

How about we display the same buttons in same order? Also I think there's a trick to add dividers, maybe with a `|` character?

image

**Comment by DeirdreYoung** [comment]
"Don't change the UI your users learned unless there is a very good reason" is a great principle.

**Comment by marph91** [comment]
Ok, let's keep the default ones as baseline, but I would like to propose two modifications:

1. Remove the "table properties" button, since it's non-functional.
2. Add the baseline and copy & paste options to the table menu. I think this is necessary, because using "ctrl + c" and "ctrl + v" for copy pasting whole lines doesn't work properly.

Or if we need to discuss it more, we could do it at the forum first.

**Comment by DeirdreYoung** [comment]
I'm just one user and this is just one opinion, but that sounds good to me.

**Comment by laurent22** [comment]
Sounds good to me too!

**Comment by marph91** [comment]
I decided to use the default table menu again (including the non-functional options). Defining the entire menu by hand is a bit more time consuming than I thought. I still think it would be good to remove some options, but maybe in another PR.

> Also I think there's a trick to add dividers, maybe with a | character?

For dividers, a pipe is used already:
https://github.com/laurent22/joplin/blob/10b9df7b9f5629726fc9e73e8e7573615c4e001d/packages/app-desktop/gui/NoteEditor/NoteBody/TinyMCE/TinyMCE.tsx#L581
I looked for other options or styling, but couldn't find anything unfortunately. It looks like the demo is using another style, since the icons are darker and there is a separator line. I guess it needs to be left open for a future PR, too.

So to summarize the PR, all needed table functionality (especially inserting rows/columns) should be available again.

**Comment by marph91** [comment]
Not sure what's wrong with the tests. They seem to fail at `@joplin/app-mobile`, which isn't touched in the PR.

**Comment by laurent22** [comment]
> Not sure what's wrong with the tests. They seem to fail at `@joplin/app-mobile`, which isn't touched in the PR.

Dev branch was not working but should be fixed now

**Comment by laurent22** [comment]
@marph91, could you provide a screenshot of the new UI?

Also there's a thread in the forum about this: https://discourse.joplinapp.org/t/table-cell-color-in-joplin-2-11-11/31554 Is there something we can do about it?

**Comment by marph91** [comment]
Right now, the top menu is same as before #8163. In the context menu, only the "table properties" are missing. I guess we should just restore the context menu as it was before #8163, if @laurent22 agrees. So what stays, is the original fix of the PR (handle the first table row as header).

By removing the "non-functional" options, I only thought of markdown notes. I wasn't aware that there are HTML notes and that some users even use these options in HTML notes.

**Comment by laurent22** [comment]
Yes please, that would be great if you could restore the orignal context menu. I agree that in theory we shouldn't have properties that don't apply to Markdown in there, but looks like people are using them anyway in pure HTML notes.

**Comment by marph91** [comment]
The original table context menu is restored now, too.

**Comment by laurent22** [comment]
Thanks @marph91, I think we can proceed with this update then
